File: core/src/main/java/com/ustadmobile/core/controller/AuditLogSelectionPresenter.java
package com.ustadmobile.core.controller;

import com.ustadmobile.core.generated.locale.MessageID;
import com.ustadmobile.core.impl.UmAccountManager;
import com.ustadmobile.core.db.UmAppDatabase;
import com.ustadmobile.core.impl.UstadMobileSystemImpl;

import java.util.ArrayList;
import java.util.Hashtable;
import java.util.LinkedHashMap;
import java.util.List;

import com.ustadmobile.core.util.UMCalendarUtil;
import com.ustadmobile.core.view.AuditLogSelectionView;
import com.ustadmobile.core.view.AuditLogListView;
import com.ustadmobile.core.view.SelectClazzesDialogView;
import com.ustadmobile.core.view.SelectMultipleTreeDialogView;
import com.ustadmobile.core.view.SelectTwoDatesDialogView;

import static com.ustadmobile.core.view.AuditLogSelectionView.ARG_AUDITLOG_ACTOR_LIST;
import static com.ustadmobile.core.view.AuditLogSelectionView.ARG_AUDITLOG_CLASS_LIST;
import static com.ustadmobile.core.view.AuditLogSelectionView.ARG_AUDITLOG_FROM_TIME;
import static com.ustadmobile.core.view.AuditLogSelectionView.ARG_AUDITLOG_LOCATION_LIST;
import static com.ustadmobile.core.view.AuditLogSelectionView.ARG_AUDITLOG_PEOPLE_LIST;
import static com.ustadmobile.core.view.AuditLogSelectionView.ARG_AUDITLOG_TO_TIME;
import static com.ustadmobile.core.view.ReportEditView.ARG_CLASSES_SET;
import static com.ustadmobile.core.view.ReportEditView.ARG_CLAZZ_LIST;
import static com.ustadmobile.core.view.ReportEditView.ARG_FROM_DATE;
import static com.ustadmobile.core.view.ReportEditView.ARG_LOCATIONS_SET;
import static com.ustadmobile.core.view.ReportEditView.ARG_LOCATION_LIST;
import static com.ustadmobile.core.view.ReportEditView.ARG_TO_DATE;


/**
 * Presenter for AuditLogSelection view
 **/
public class AuditLogSelectionPresenter extends UstadBaseController<AuditLogSelectionView> {

    UmAppDatabase repository;

    private List<Long> selectedClasses;
    private List<Long> selectedLocations;
    private List<Long> selectedPeople;
    private List<Long> selectedActors;

    private long fromTime, toTime;

    private static final int TIME_PERIOD_LAST_WEEK = 1;
    private static final int TIME_PERIOD_LAST_TWO_WEEK = 2;
    private static final int TIME_PERIOD_LAST_MONTH = 3;
    private static final int TIME_PERIOD_LAST_THREE_MONTHS = 4;
    private static final int TIME_PERIOD_CUSTOM = 5;


    public AuditLogSelectionPresenter(Object context, Hashtable arguments, AuditLogSelectionView view) {
        super(context, arguments, view);

        repository = UmAccountManager.getRepositoryForActiveAccount(context);
    }

    /**
     * Generates the time period options. Generates the hashmap and sends it to the view.
     */
    private void updateTimePeriod(){
        LinkedHashMap<Integer, String> timePeriodOptions = new LinkedHashMap<>();
        UstadMobileSystemImpl impl = UstadMobileSystemImpl.getInstance();
        timePeriodOptions.put(TIME_PERIOD_LAST_WEEK,
                impl.getString(MessageID.last_week, context));
        timePeriodOptions.put(TIME_PERIOD_LAST_TWO_WEEK,
                impl.getString(MessageID.last_two_weeks, context));
        timePeriodOptions.put(TIME_PERIOD_LAST_MONTH,
                impl.getString(MessageID.last_month, context));
        timePeriodOptions.put(TIME_PERIOD_LAST_THREE_MONTHS,
                impl.getString(MessageID.last_three_months, context));
        timePeriodOptions.put(TIME_PERIOD_CUSTOM,
                impl.getString(MessageID.custom_date_range, context));

        view.populateTimePeriod(timePeriodOptions);

    }

    @Override
    public void onCreate(Hashtable savedState) {
        super.onCreate(savedState);

        //Update time period options.
        updateTimePeriod();

    }

    public void handleClickPrimaryActionButton() {

        UstadMobileSystemImpl impl = UstadMobileSystemImpl.getInstance();
        Hashtable args = new Hashtable();
        args.put(ARG_AUDITLOG_FROM_TIME, fromTime);
        args.put(ARG_AUDITLOG_TO_TIME, toTime);

        if(selectedClasses != null && !selectedClasses.isEmpty()){
            Long[] classesArray = new Long[selectedClasses.size()];
            selectedClasses.toArray(classesArray);
            args.put(ARG_AUDITLOG_CLASS_LIST, classesArray);
        }

        if(selectedLocations != null && !selectedLocations.isEmpty()){
            Long[] locationsArray = new Long[selectedLocations.size()];
            selectedLocations.toArray(locationsArray);
            args.put(ARG_AUDITLOG_LOCATION_LIST, locationsArray);
        }

        if(selectedPeople != null && !selectedPeople.isEmpty()){
            Long[] peopleArray = new Long[selectedPeople.size()];
            selectedPeople.toArray(peopleArray);
            args.put(ARG_AUDITLOG_PEOPLE_LIST, peopleArray);
        }

        if(selectedActors != null && !selectedActors.isEmpty()){
            Long[] actorArray = new Long[selectedActors.size()];
            selectedActors.toArray(actorArray);
            args.put(ARG_AUDITLOG_ACTOR_LIST, actorArray);
        }

        impl.go(AuditLogListView.VIEW_NAME, args, context);
    }


    public void handleTimePeriodSelected(int selected) {
        toTime = System.currentTimeMillis();
        selected++;
        switch (selected){
            case TIME_PERIOD_LAST_WEEK:
                fromTime = UMCalendarUtil.getDateInMilliPlusDays(-7);
                break;
            case TIME_PERIOD_LAST_TWO_WEEK:
                fromTime = UMCalendarUtil.getDateInMilliPlusDays(-14);
                break;
            case TIME_PERIOD_LAST_MONTH:
                fromTime = UMCalendarUtil.getDateInMilliPlusDays(-31);
                break;
            case TIME_PERIOD_LAST_THREE_MONTHS:
                fromTime = UMCalendarUtil.getDateInMilliPlusDays(-61);
                break;
            case TIME_PERIOD_CUSTOM:
                toTime = 0L;
                goToSelectTwoDatesDialog();
                break;
        }
    }

    private void goToSelectTwoDatesDialog(){
        UstadMobileSystemImpl impl = UstadMobileSystemImpl.getInstance();
        Hashtable args = new Hashtable();
        impl.go(SelectTwoDatesDialogView.VIEW_NAME, args, context);
    }

    public void goToSelectClassesDialog() {
        UstadMobileSystemImpl impl = UstadMobileSystemImpl.getInstance();
        Hashtable<String, Object> args = new Hashtable<>();


        if(selectedLocations != null && !selectedLocations.isEmpty()){
            Long[] selectedLocationsArray =
                    ReportOverallAttendancePresenter.convertLongList(selectedLocations);
            args.put(ARG_LOCATIONS_SET, selectedLocationsArray);
        }

        if(selectedClasses != null && !selectedClasses.isEmpty()) {
            Long[] selectedClassesArray =
                    ReportOverallAttendancePresenter.convertLongList(selectedClasses);
            args.put(ARG_CLASSES_SET, selectedClassesArray);
        }
        impl.go(SelectClazzesDialogView.VIEW_NAME, args, context);
    }

    public void goToLocationDialog() {
        UstadMobileSystemImpl impl = UstadMobileSystemImpl.getInstance();
        Hashtable<String, Object> args = new Hashtable<>();

        if(selectedLocations != null && !selectedLocations.isEmpty()){
            Long[] selectedLocationsArray =
                    ReportOverallAttendancePresenter.convertLongList(selectedLocations);
            args.put(ARG_LOCATIONS_SET, selectedLocationsArray);
        }

        impl.go(SelectMultipleTreeDialogView.VIEW_NAME, args, context);
    }

    public void goToPersonDialog() {

        UstadMobileSystemImpl impl = UstadMobileSystemImpl.getInstance();
        Hashtable<String, Object> args = new Hashtable<>();

        if(selectedPeople != null && !selectedPeople.isEmpty()){
            Long[] selectedPeopleArray =
                    ReportOverallAttendancePresenter.convertLongList(selectedPeople);
            args.put(ARG_LOCATIONS_SET, selectedPeopleArray);
        }

        //TODO: This
        //impl.go(SelectPeopleDialogView.VIEW_NAME, args, context);
    }

    public void goToActorDialog() {

        UstadMobileSystemImpl impl = UstadMobileSystemImpl.getInstance();
        Hashtable<String, Object> args = new Hashtable<>();

        if(selectedActors != null && !selectedActors.isEmpty()){
            Long[] selectedPeopleArray =
                    ReportOverallAttendancePresenter.convertLongList(selectedActors);
            args.put(ARG_LOCATIONS_SET, selectedPeopleArray);
        }

        //TODO: This
        //impl.go(SelectPeopleDialogView.VIEW_NAME, args, context);
    }


    //Getters and Setters:

    public List<Long> getSelectedClasses() {
        return selectedClasses;
    }

    public void setSelectedClasses(List<Long> selectedClasses) {
        if(selectedClasses == null){
            selectedClasses = new ArrayList<>();
        }
        this.selectedClasses = selectedClasses;
    }

    public List<Long> getSelectedLocations() {
        return selectedLocations;
    }

    public void setSelectedLocations(List<Long> selectedLocations) {
        if(selectedLocations == null){
            selectedLocations = new ArrayList<>();
        }
        this.selectedLocations = selectedLocations;
    }

    public List<Long> getSelectedPeople() {
        return selectedPeople;
    }

    public void setSelectedPeople(List<Long> selectedPeople) {
        if(selectedPeople == null){
            selectedPeople = new ArrayList<>();
        }
        this.selectedPeople = selectedPeople;
    }

    public List<Long> getSelectedActors() {
        return selectedActors;
    }

    public void setSelectedActors(List<Long> selectedActors) {
        if(selectedActors == null){
            selectedActors = new ArrayList<>();
        }
        this.selectedActors = selectedActors;
    }

    public long getFromTime() {
        return fromTime;
    }

    public void setFromTime(long fromTime) {
        this.fromTime = fromTime;
    }

    public long getToTime() {
        return toTime;
    }

    public void setToTime(long toTime) {
        this.toTime = toTime;
    }



}

File: core/src/commonMain/kotlin/com/ustadmobile/staging/core/controller/AuditLogSelectionPresenter.kt
package com.ustadmobile.core.controller

import com.ustadmobile.core.db.UmAppDatabase
import com.ustadmobile.core.generated.locale.MessageID
import com.ustadmobile.core.impl.UmAccountManager
import com.ustadmobile.core.impl.UstadMobileSystemImpl
import com.ustadmobile.core.util.ArgumentUtil
import com.ustadmobile.core.util.UMCalendarUtil
import com.ustadmobile.core.view.*
import com.ustadmobile.core.view.AuditLogSelectionView.Companion.ARG_AUDITLOG_ACTOR_LIST
import com.ustadmobile.core.view.AuditLogSelectionView.Companion.ARG_AUDITLOG_CLASS_LIST
import com.ustadmobile.core.view.AuditLogSelectionView.Companion.ARG_AUDITLOG_FROM_TIME
import com.ustadmobile.core.view.AuditLogSelectionView.Companion.ARG_AUDITLOG_LOCATION_LIST
import com.ustadmobile.core.view.AuditLogSelectionView.Companion.ARG_AUDITLOG_PEOPLE_LIST
import com.ustadmobile.core.view.AuditLogSelectionView.Companion.ARG_AUDITLOG_TO_TIME
import com.ustadmobile.core.view.ReportEditView.Companion.ARG_ACTOR_SET
import com.ustadmobile.core.view.ReportEditView.Companion.ARG_CLASSES_SET
import com.ustadmobile.core.view.ReportEditView.Companion.ARG_LOCATIONS_SET
import com.ustadmobile.core.view.ReportEditView.Companion.ARG_PEOPLE_SET


/**
 * Presenter for AuditLogSelection view
 */
class AuditLogSelectionPresenter(context: Any, arguments: Map<String, String>?, view:
AuditLogSelectionView, val impl : UstadMobileSystemImpl = UstadMobileSystemImpl.instance)
    : UstadBaseController<AuditLogSelectionView>(context, arguments!!, view) {

    internal var repository: UmAppDatabase

    //Getters and Setters:

    var selectedClasses: List<Long>? = null
        set(selectedClasses) {
            var selectedClasses = selectedClasses
            if (selectedClasses == null) {
                selectedClasses = ArrayList()
            }
            field = selectedClasses
        }
    var selectedLocations: List<Long>? = null
        set(selectedLocations) {
            var selectedLocations = selectedLocations
            if (selectedLocations == null) {
                selectedLocations = ArrayList()
            }
            field = selectedLocations
        }
    var selectedPeople: List<Long>? = null
        set(selectedPeople) {
            var selectedPeople = selectedPeople
            if (selectedPeople == null) {
                selectedPeople = ArrayList()
            }
            field = selectedPeople
        }
    var selectedActors: List<Long>? = null
        set(selectedActors) {
            var selectedActors = selectedActors
            if (selectedActors == null) {
                selectedActors = ArrayList()
            }
            field = selectedActors
        }

    var fromTime: Long = 0
    var toTime: Long = 0


    init {

        repository = UmAccountManager.getRepositoryForActiveAccount(context)
    }

    /**
     * Generates the time period options. Generates the hashmap and sends it to the view.
     */
    private fun updateTimePeriod() {
        val timePeriodOptions = HashMap<Int, String>()
        timePeriodOptions[TIME_PERIOD_LAST_WEEK] = impl.getString(MessageID.last_week, context)
        timePeriodOptions[TIME_PERIOD_LAST_TWO_WEEK] = impl.getString(MessageID.last_two_weeks, context)
        timePeriodOptions[TIME_PERIOD_LAST_MONTH] = impl.getString(MessageID.last_month, context)
        timePeriodOptions[TIME_PERIOD_LAST_THREE_MONTHS] = impl.getString(MessageID.last_three_months, context)
        timePeriodOptions[TIME_PERIOD_CUSTOM] = impl.getString(MessageID.custom_date_range, context)

        view.populateTimePeriod(timePeriodOptions)

    }

    override fun onCreate(savedState: Map<String, String>?) {
        super.onCreate(savedState)

        //Update time period options.
        updateTimePeriod()

    }

    fun handleClickPrimaryActionButton() {

        val args = HashMap<String, String>()
        args.put(ARG_AUDITLOG_FROM_TIME, fromTime.toString())
        args.put(ARG_AUDITLOG_TO_TIME, toTime.toString())

        if (this.selectedClasses != null && !this.selectedClasses!!.isEmpty()) {
            val classesArray = arrayOfNulls<Long>(this.selectedClasses!!.size)
            this.selectedClasses!!.toTypedArray()
            //TODOne: KMP flatten out list to CSVs
            val classesCSV = ArgumentUtil.convertLongListToStringCSV(selectedClasses!!)
            args.put(ARG_AUDITLOG_CLASS_LIST, classesCSV)
        }

        if (this.selectedLocations != null && !this.selectedLocations!!.isEmpty()) {
            val locationsArray = arrayOfNulls<Long>(this.selectedLocations!!.size)
            this.selectedLocations!!.toTypedArray()
            //TODOne: KMP flatten out list to CSVs
            val locationsCSV = ArgumentUtil.convertLongListToStringCSV(selectedLocations!!)
            args.put(ARG_AUDITLOG_LOCATION_LIST, locationsCSV)
        }

        if (this.selectedPeople != null && !this.selectedPeople!!.isEmpty()) {
            val peopleArray = arrayOfNulls<Long>(this.selectedPeople!!.size)
            this.selectedPeople!!.toTypedArray()
            //TODOne: KMP flatten out list to CSVs
            val peopleCSV = ArgumentUtil.convertLongListToStringCSV(selectedPeople!!)
            args.put(ARG_AUDITLOG_PEOPLE_LIST, peopleCSV)
        }

        if (this.selectedActors != null && !this.selectedActors!!.isEmpty()) {
            val actorArray = arrayOfNulls<Long>(this.selectedActors!!.size)
            this.selectedActors!!.toTypedArray()
            //TODOne: KMP flatten out list to CSVs
            val actorCSV = ArgumentUtil.convertLongListToStringCSV(selectedActors!!)
            args.put(ARG_AUDITLOG_ACTOR_LIST, actorCSV)
        }

        impl.go(AuditLogListView.VIEW_NAME, args, context)
    }


    fun handleTimePeriodSelected(selected: Int) {
        var selected = selected
        toTime = UMCalendarUtil.getDateInMilliPlusDays(0)
        selected++
        when (selected) {
            TIME_PERIOD_LAST_WEEK -> fromTime = UMCalendarUtil.getDateInMilliPlusDays(-7)
            TIME_PERIOD_LAST_TWO_WEEK -> fromTime = UMCalendarUtil.getDateInMilliPlusDays(-14)
            TIME_PERIOD_LAST_MONTH -> fromTime = UMCalendarUtil.getDateInMilliPlusDays(-31)
            TIME_PERIOD_LAST_THREE_MONTHS -> fromTime = UMCalendarUtil.getDateInMilliPlusDays(-61)
            TIME_PERIOD_CUSTOM -> {
                toTime = 0L
                goToSelectTwoDatesDialog()
            }
        }
    }

    private fun goToSelectTwoDatesDialog() {
        val args = HashMap<String, String>()
        impl.go(SelectTwoDatesDialogView.VIEW_NAME, args, context)
    }

    fun goToSelectClassesDialog() {
        val args = HashMap<String, String>()


        if (this.selectedLocations != null && !this.selectedLocations!!.isEmpty()) {
            val selectedLocationsArray = ReportOverallAttendancePresenter.convertLongList(this.selectedLocations!!)
            //TODOne: KMP flatten out list to CSVs
            val selectedLocationsCSV = ArgumentUtil.convertLongListToStringCSV(selectedLocations!!)
            args.put(ARG_LOCATIONS_SET, selectedLocationsCSV)
        }

        if (this.selectedClasses != null && !this.selectedClasses!!.isEmpty()) {
            val selectedClassesArray = ReportOverallAttendancePresenter.convertLongList(this.selectedClasses!!)
            //TODOne: KMP flatten out list to CSVs
            val selectedClassesCSV = ArgumentUtil.convertLongListToStringCSV(selectedClasses!!)
            args.put(ARG_CLASSES_SET, selectedClassesCSV)
        }
        //impl.go(SelectClazzesDialogView.VIEW_NAME, args, context)
    }

    fun goToLocationDialog() {
        val args = HashMap<String, String>()

        if (this.selectedLocations != null && !this.selectedLocations!!.isEmpty()) {
            val selectedLocationsArray = ReportOverallAttendancePresenter.convertLongList(this.selectedLocations!!)
            //TODOne: KMP flatten out list to CSVs
            val selectedLocationsCSV = ArgumentUtil.convertLongListToStringCSV(selectedLocations!!)
            args.put(ARG_LOCATIONS_SET, selectedLocationsCSV)
        }

        impl.go(SelectMultipleTreeDialogView.VIEW_NAME, args, context)
    }

    fun goToPersonDialog() {

        val args = HashMap<String, String>()

        if (this.selectedPeople != null && !this.selectedPeople!!.isEmpty()) {
            val selectedPeopleArray = ReportOverallAttendancePresenter.convertLongList(this.selectedPeople!!)
            //TODOne: KMP flatten out list to CSVs
            val selectedPeopleCSV = ArgumentUtil.convertLongListToStringCSV(selectedPeople!!)
            args.put(ARG_PEOPLE_SET, selectedPeopleCSV)
        }

        //impl.go(SelectPeopleDialogView.VIEW_NAME, args, context)
    }

    fun goToActorDialog() {

        val args = HashMap<String, String>()

        if (this.selectedActors != null && !this.selectedActors!!.isEmpty()) {
            val selectedPeopleArray = ReportOverallAttendancePresenter.convertLongList(this.selectedActors!!)
            //TODOne: KMP flatten out list to CSVs
            val selectedActorCSV = ArgumentUtil.convertLongListToStringCSV(selectedActors!!)
            args.put(ARG_ACTOR_SET, selectedActorCSV)
        }
        //args.put(ARG_SELECT_ACTOR, "yes")
        //impl.go(SelectPeopleDialogView.VIEW_NAME, args, context)
    }

    companion object {

        private val TIME_PERIOD_LAST_WEEK = 1
        private val TIME_PERIOD_LAST_TWO_WEEK = 2
        private val TIME_PERIOD_LAST_MONTH = 3
        private val TIME_PERIOD_LAST_THREE_MONTHS = 4
        private val TIME_PERIOD_CUSTOM = 5
    }


}
